ClosureCompilerRecursive.java

```diff
@@ -0,0 +1,123 @@
+import java.io.IOException;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.PrintWriter;
+import java.io.File;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.stream.Stream;
+
+import com.google.javascript.jscomp.CompilationLevel;
+import com.google.javascript.jscomp.Compiler;
+import com.google.javascript.jscomp.CompilerOptions;
+import com.google.javascript.jscomp.SourceFile;
+import com.google.javascript.jscomp.CommandLineRunner;
+import com.google.javascript.jscomp.CompilerOptions.LanguageMode;
+
+public class ClosureCompilerRecursive {
+   protected Path directory;
+   protected Integer errorCount = 0;
+   protected Boolean writeFiles = true;
+
+   protected String compile(File file) {
+      Compiler compiler = new Compiler();
+
+      List<SourceFile> list = null;
+
+      try {
+          list =
+         CommandLineRunner
+         .getBuiltinExterns(CompilerOptions.Environment.BROWSER);
```

Comment:
What does this do?

Reply:
Tells CC which global variables are already defined. I wasn't 100% certain why that was needed until I just read this: https://dianjin.github.io/introduction-to-closure-externs/

Sounds like it's only (most?) useful when using the Advanced mode. The advanced mode renames everything it can (even global vars / functions / class names). But it needs to know which things to leave alone (browser APIs).

We don't use advanced mode.

Reply:
That would make a fantastic comment ^